### DISCUSSION TITLE
POC of perpetual markets on Miden

### DISCUSSION THREAD

**Discussion by armanthepythonguy** [0xPolygonMiden/examples#208] [Ideas]
@Dominik1999 As we discussed last week, I have some proposals for a private perpetual market on Miden.
Leveraging the privacy-preserving capabilities of the Miden blockchain, our project will build a private perpetual market where users can trade assets without exposing their transaction details and balances. Instead of just going with the single future market approach as we can use Miden VM to have complex computation we would also like to develop a perpetual Options market on Miden.

**Private Leveraged Index Perpetuals:**

For the private leveraged perpetual markets, we propose to create our own indices which can have multiple asset classes bagged together. We can have these markets rebalanced privately.

1. Users can take positions on indices privately, with leverage applied homomorphically.
2. As the underlying assets of the index change in value, the index is rebalanced in a private manner. Users’ positions are adjusted based on the new index composition, without revealing the underlying assets or weights. Traders can go long or short on an index in perpetuity, adjusting their leverage or rebalancing their position without the need for settlement.
3. Dynamic Funding Rates

**Private Perpetual Options Markets**

Normal blockchains face difficulty in handling options trade because of the computational complexity the options market holds. Using Miden VM, we can create a private perpetual options market with the futures market. 

1. Create a system where users can hedge their positions with perpetual options contracts, shifting between calls and puts based on market conditions. Both options and underlying assets remain private.
2. Liquidations occur privately if collateral requirements aren’t met, maintaining the security of the options market without revealing sensitive data.
3. Dynamic Funding Rates

Also, I saw @phklive building a central limit order book. Maybe he can help us brainstorm some more ideas.

**Comment by phklive** [comment]
I am indeed building an order-book on top of Miden. 

Would be glad to help you on your journey.

My telegram is the same as my github, feel free to ping.

**Comment by Dominik1999** [comment]
That sounds great! Happy to help you guys to get started!

Can you ping me again on Telegram and we can see how we sort things out.

**Comment by armanthepythonguy** [comment]
# Private Perpetual Market

## Vision
 - Traders can open leveraged positions on various assets with minimal gas fees and robust privacy guarantees.
 - Liquidity Providers can supply collateral to fund leveraged positions in return for interest or a share of transaction fees.
 - Automated Liquidators maintain healthy collateral ratios in the market, safeguarding the system against under-collateralized positions.

## Financial Mechanics and Leveraged Position Examples

###  Interest Rates on Borrowed Capital
 - Purpose: LPs lend assets to traders, who use them to amplify positions. The interest rate on borrowed capital compensates LPs for their risk, providing an incentive to supply liquidity.
 - Variable or Fixed Rates: Rates can be variable, depending on the demand for leverage, or set by the governance. Variable rates ensure that high demand for leverage yields higher returns for LPs, while low demand decreases costs for traders.
 - A small trading fee on each position is charged to traders and split between liquidity providers, governance participants, and potentially a treasury for future upgrades.

 ### Funding Rate
Funding rates balance the perpetual market by encouraging long or short positions to converge around the underlying asset's spot price. These rates are charged or paid periodically depending on the open position’s direction relative to market demand, creating a peer-to-peer adjustment to align the position costs with market conditions.

 - ### Example of Funding Rate in Action:

  - If long positions exceed short positions by a significant margin, the funding rate will be positive, meaning long traders pay short traders.
  - Conversely, if short positions are greater, the funding rate is negative, so short traders pay long traders.

### Trade Examples
 - Example 1: Profit Scenario
   - Suppose Alice opens a 5x leveraged long position on an asset priced at $100. She invests $1,000, which effectively gives her exposure to $5,000 (5x leverage).
Price Increase: The asset price rises by 10% to $110.
Position Value: Alice’s position is now worth $5,500.
Profit: Alice’s initial investment of $1,000 has generated a profit of $500 (50% gain on her $1,000).
 - Example 2: Loss Scenario
   - Alice has a 5x leveraged long position on an asset priced at $100. She has invested $1,000, meaning her position value is $5,000.
Price Decline: The asset price drops by 20% to $80.
Position Value: Her position is now worth $4,000.
Collateral Ratio Check: With 5x leverage, a 20% drop means that her collateral no longer covers her leveraged exposure. The protocol triggers liquidation if her collateral drops below a certain threshold, say 80% of her initial collateral.
Liquidation: Automated liquidators step in to close Alice’s position to prevent further loss. They receive a portion of her collateral as a liquidation penalty, which compensates them for protecting the protocol’s solvency.
